File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        
        vector<int> ans; 
        dfs(root ,0,ans); 
        return ans; 

    }
        
void dfs(TreeNode* root, int dep, vector<int> &ans){
    if(!root) return ; 
    if(dep >= ans.size()) ans.push_back(root->val); 
    dfs(root->right , dep+1,ans); 
    dfs(root->left , dep+1 ,ans); 


        
    }
};

File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        
        if not root: return 
        dic = defaultdict(list)
        q = deque()
        q.append([root,0])
        
        while q: 
            node,dep = q.popleft() 
            dic[dep].append(node.val)
            
            if node.left: 
                q.append([node.left ,dep+1])
            
            if node.right:
                q.append([node.right, dep+1])
        
        return [ v[-1] for k,v in dic.items() ] 